W-11986472: add documentation for syntax directive and compatibility … (docs W-12272475) (#214)
mulesoft/docs-dataweave: modules/ROOT/pages/dataweave-versioning-behavior.adoc

= Versioning Behavior in DataWeave

> In each version of DataWeave, changes are introduced that modify the way the language behaves. When existing scripts depend on previous behavior, system properties are introduced to toggle the change so that backwards compatibilty is not broken in the current version. These system properties that control behavior in DataWeave are called *Compatibility Flags*. Each flag defined a deletion version, after which it will no longer exists, and the behavior that it controlled will default to the new one.

Reviewer: ```suggestion
New minor versions of DataWeave can introduce changes that modify the way the language behaves. Starting in DataWeave 2.5, you can use DataWeave system properties (compatibility flags) to continue using existing scripts that depend on the previous behavior. However, each flag defines a deletion version, which identifies the version in which the flag is no longer available and the behavior that it controls defaults to the changed behavior.
```

Author: The only thing that I don't know if it's clear from this paragraph is that all compatibility flags are system properties but not all system properties are compatibility flags. We still have properties for different memory/cpu settings that don't have deletion version.

Reviewer: @andres-rad Sounds like we need to identify which are which, probably with a deletion version: https://docs.mulesoft.com/dataweave/2.4/dataweave-system-properties 

We also say that some experimental ones are subject to deletion at a future date. 

We should probably something like this to the description and point to the new version doc for more info about how to retain the old version's behavior. 

* Removal version: 2.4

Author: I think that this change is good, you can merge if you think the newly autogenerated system properties are ok.

Reviewer: @andres-rad Thx, can you commit it so that I can avoid an SoD violation.

Reviewer: I will check the autogenerated properties separately. I have a ToDo for that.

Reviewer: To self: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001IFA61YAH/view for checking this info in autogenerated content.

> Specifying DataWeave's *language level* indicates to the runtime which Compatibility Flags are available. As previously stated, when upgrading to a new version there are flags that will be deleted. In order to facilitate updatating DataWeave, setting the language level of a previous version makes it so that the flags available on that version are visible, mantaining backwards compatibility. Language level defaults to the running DataWeave version and has to be at most the same as the runtime version.

Reviewer: ```suggestion
Specifying a DataWeave language level indicates to the DataWeave runtime what compatibility flags are available. To upgrade and maintain backward compatibility, you can set the language level of a previous version so that the flags available on that version remain visible even if they expire in the DataWeave version to which you upgrade. The language level defaults to the running DataWeave version and has to be at most the same as the runtime version.
```

Reviewer: @svacas What does "and has to be at most the same as the runtime version" mean? Do you mean that the DW version cannot exceed the DataWeave runtime version that you are running? Or what? Thx!

Author: It means the DataWeave runtime version running. In a mule app it is the DW that is packaged in the application.

> For mule applications the language level is determined by the app's *minMuleVersion*. Given the minMuleVersion, the language level is computed by substracting 2 major versions as shown in the following table:

Reviewer: ```suggestion
For Mule applications, the language level is determined by the application's `minMuleVersion`. Given the `minMuleVersion`, the language level is computed by subtracting two major versions, as the following table shows:
```

> For Mule applications, the language level is determined by the application's `minMuleVersion`. Given the `minMuleVersion`, the language level is computed by subtracting two major versions, as the following table shows:

Reviewer: ```suggestion
For Mule applications, the language level is determined by the application's `minMuleVersion` setting (see xref:mule-runtime::feature-flagging.adoc#feature-flags-reference[Feature Flags Reference] in the Mule documentation). Given the `minMuleVersion`, the language level is computed by subtracting two major versions, as the following table shows:
```

[%header%autowidth.spread,cols="a,a"]
|===

> | minMuleVersion | LanguageLevel

Reviewer: ```suggestion
| minMuleVersion | Language Level
```

| 4.5 | 2.5
| 4.4 | 2.4
| 4.3 | 2.3
|===

> 

Reviewer: ```suggestion

Setting the Mule version configures the application to use the corresponding DataWeave language level. This setting is independent of the language level set in a DataWeave script's `%dw` directive. For details, see xref:dataweave-language-introduction.adoc#dw_header[DataWeave Header] and xref:dataweave-versioning-behavior.adoc[].
```

Author: I would say that the setting is independent of the syntax version set in the script by the directive.

Reviewer: revised, thx @andres-rad !

> 

Reviewer: ```suggestion

Setting the Mule version configures the application to use the corresponding DataWeave language level. This setting is independent of the language level set in a DataWeave script's `%dw` directive. For details, see xref:dataweave-language-introduction.adoc#dw_header[DataWeave Header] and xref:dataweave-versioning-behavior.adoc[].
```

> For example, the `com.mulesoft.dw.xml_reader.honourMixedContentStructure` exists for language level 2.4 (it was introduced on that version and is deleted in 2.5), meaning that for DataWeave 2.5 the property will be visible only if the language level is set to 2.4, otherwise it will default to its new behavior.

Reviewer: ```suggestion
For example, the `com.mulesoft.dw.xml_reader.honourMixedContentStructure` system property exists for language level 2.4, when it was introduced, and it is removed in DataWeave 2.5. In DataWeave 2.5, the property will be visible only if you set the language level to 2.4. Otherwise, the property will default to its new behavior in 2.5.
```

Reviewer: You would set minMuleVersion to 2.4 or set the %dw directive to that?

Author: Yeah, we should call that a compatibility flag.

The minMuleVersion is always set on the mule app and is a property of the whole application it's not just related to DW. Regarding the directive they are unrelated features, so for example if you want to migrate a mule app from 4.4.0 to 4.5.0 you can set the app language level to 2.4 so your DW app doesn't break, but you can use the `%dw 2.5` in your scripts, as 4.5.0 comes with DW 2.5.

Let me know is this clears things up

Reviewer: That clarifies it for me. Thanks!
